server/server.go

```diff
@@ -6,53 +6,299 @@ package server
+	polling       polling
+	notifications notifications
+
+	close *sync.Once
 
 	pb.UnimplementedProducerServer
 }
 
-// PublishEvents is the server implementation of the gRPC PublishEvents call
-func (serv shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
-	results := []*messages.EventResult{}
-	for _, evt := range req.Events {
-		serv.logger.Infof("Got event %s: %#v", evt.EventId, evt.Fields.Data)
-		err := serv.queue.Publish(evt)
-		if err != nil {
-			// If we couldn't accept any events, return the error directly. Otherwise,
-			// just return success on however many events we were able to handle.
-			if len(results) == 0 {
-				return nil, err
+type polling struct {
+	ctx     context.Context
+	stop    func()
+	stopped chan struct{}
+}
+
+// NewShipperServer creates a new server instance for handling gRPC endpoints.
+func NewShipperServer(q *queue.Queue, cfg ShipperServerConfig) (ShipperServer, error) {
+	if q == nil {
+		return nil, errors.New("queue cannot be nil")
+	}
+
+	err := cfg.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
+	id, err := uuid.NewV4()
+	if err != nil {
+		return nil, err
+	}
+
+	s := shipperServer{
+		uuid:   id.String(),
+		logger: logp.NewLogger("shipper-server"),
+		queue:  q,
+		cfg:    cfg,
+		polling: polling{
+			stopped: make(chan struct{}),
+		},
+		notifications: notifications{
+			subscribers: make(map[uuid.UUID]chan change),
+			mutex:       &sync.Mutex{},
+		},
+		close: &sync.Once{},
+	}
+
+	s.polling.ctx, s.polling.stop = context.WithCancel(context.Background())
+	s.startPolling()
+
+	return &s, nil
+}
+
+// GetAcceptedIndex atomically reads the accepted index
+func (serv *shipperServer) GetAcceptedIndex() int64 {
+	return atomic.LoadInt64(&serv.acceptedIndex)
+}
+
+// GetPersistedIndex atomically reads the persisted index
+func (serv *shipperServer) GetPersistedIndex() int64 {
+	return atomic.LoadInt64(&serv.persistedIndex)
+}
+
+// PublishEvents is the server implementation of the gRPC PublishEvents call.
+func (serv *shipperServer) PublishEvents(_ context.Context, req *messages.PublishRequest) (*messages.PublishReply, error) {
+	resp := &messages.PublishReply{
+		Uuid: serv.uuid,
+	}
+
+	// the value in the request is optional
+	if req.Uuid != "" && req.Uuid != serv.uuid {
+		resp.AcceptedIndex = serv.GetAcceptedIndex()
+		resp.PersistedIndex = serv.GetPersistedIndex()
+		serv.logger.
+			With(
+				"expected", serv.uuid,
+				"actual", req.Uuid,
+			).
+			Debugf("shipper UUID does not match, all events rejected")
+
+		return resp, nil
+	}
+
+	for _, e := range req.Events {
+		err := serv.queue.Publish(e)
+		if err == nil {
+			resp.AcceptedIndex++
+			continue
+		}
+
+		log := serv.logger.
+			With(
+				"event_count", len(req.Events),
+				"accepted_count", resp.AcceptedCount,
+			)
+
+		if errors.Is(err, queue.ErrQueueIsFull) {
+			log.Warnf("queue is full, not all events accepted")
```

Comment:
This could be a very noisy log message, I think it might be better to track this error and the "failed to enqueue" error with a metrics counter. 

We may need to setup the internal monitoring metrics for the shipper first though, I'm not sure if anyone has done that yet.

Reply:
I will demote it to the debug level. Let's address monitoring in a future PR.

Reply:
Demoted in https://github.com/elastic/elastic-agent-shipper/pull/76/commits/8665574ea664eb4ce019a1085909db85125eb074

Reply:
https://github.com/elastic/elastic-agent-shipper/issues/78 for adding monitoring/metrics